chainer/function_node.py

```diff
@@ -0,0 +1,591 @@
+            Tuple of output arrays. Each element can be NumPy or CuPy arrays.
+
+        .. warning::
+
+            Implementation of :class:`FunctionNode` must take care that the
+            return value must be a tuple even if it returns only one array.
+
+        """
+        raise NotImplementedError
+
+    def forward_gpu(self, inputs):
+        """Computes the output arrays from the input CuPy arrays.
+
+        Args:
+            inputs: Tuple of input :class:`cupy.ndarray` objects.
+
+        Returns:
+            Tuple of output arrays. Each element can be NumPy or CuPy arrays.
+
+        .. warning::
+
+            Implementation of :class:`FunctionNode` must take care that the
+            return value must be a tuple even if it returns only one array.
+
+        """
+        raise NotImplementedError
+
+    def retain_inputs(self, indexes):
+        """Lets specified input variable nodes keep data arrays.
+
+        By calling this method from :meth:`forward`, the function node can
+        specify which inputs are required for backprop. The input variables
+        with retained arrays can be obtained by :meth:`get_retained_inputs`
+        from :meth:`backward`.
+
+        Unlike :class:`Function`, the function node **DOES NOT** keep input
+        arrays by default. If you want to keep some or all input arrays, do not
+        forget to call this method.
+
+        Note that **this method must not be called from the outside of
+        forward method.**
+
+        Args:
+            indexes (iterable of int): Indexes of input variables that the
+                function does not require for backprop.
+
+        """
+        self._input_indexes_to_retain = indexes
+
+    def retain_outputs(self, indexes):
+        """Lets specified output variable nodes keep data arrays.
+
+        By calling this method from :meth:`forward`, the function node can
+        specify which outputs are required for backprop. If this method is not
+        called, any output variables are not marked to keep the data array at
+        the point of returning from :meth:`apply`. The output variables with
+        retained arrays can be obtained by :meth:`get_retained_outputs` from
+        :meth:`backward`.
+
+        .. note::
+
+           It is recommended to use this method if the function requires some
+           or all output arrays in backprop. The function can also use output
+           arrays just by keeping references to them directly, whereas it might
+           influence on the performance of later function applications to the
+           output variables.
+
+        Note that **this method must not be called from the outside of
+        forward method.**
+
+        Args:
+            indexes (iterable of int): Indexes of input variables that the
+                function does not require for backprop.
+
+        """
+        self._output_indexes_to_retain = indexes
+
+    def backward(self, target_input_indexes, grad_outputs):
```

Comment:
Hmm, do you come up with a better name? If we rename it, I think we should rename `forward` as well for consistency (I mean the names of `forward` and `backward` should have a clear correspondence).